README.md
[![MBARI](https://www.mbari.org/wp-content/uploads/2014/11/logo-mbari-3b.png)](http://www.mbari.org)

[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
![Supported Platforms](https://img.shields.io/badge/Supported%20Platforms-Windows%20%7C%20macOS%20%7C%20Linux-green)
![license-GPL](https://img.shields.io/badge/license-GPL-blue)

# About

*deepsea-track* is online tracking software for tracking multiple objects in deep sea underwater video.
It is designed to be used with output from object detection models sent over [ZeroMQ](https://zeromq.org/)
 or in data formatted into XML files in [PASCAL VOC](http://host.robots.ox.ac.uk/pascal/VOC/) format. 
It generates unique *VisualEvent* track sequences in JSON format.

---

[click image below to see example in YouTube]
[![Example video output](https://img.youtube.com/vi/cMZ8vr0aAYI/maxresdefault.jpg)](https://youtu.be/cMZ8vr0aAYI)

# Questions?

If you have any questions or are interested in contributing, please contact me at dcline@mbari.org.

*Danelle Cline*
https://www.mbari.org/cline-danelle-e/

---

# Requirements
- [Docker](www.docker.com)

Alternatively, can be built natively for Mac with

- A compiler that support >= C++11
- [CMake `>= 3.15`](https://cmake.org/download/)
- [HomeBrew](https://brew.sh/) for Mac OS install

---

## How to use deepsea-track

* Create a folder, e.g. "benthic" and add your video, class map, configuration file and folder
 to store the results, e.g.:
        
~~~
│   └── benthic
│       ├── video.mp4
│       ├── deepsea_class_map.json
│       ├── deepsea_cfg.json
│       ├── f001000.json
│       ├── f001001.json
│       ├── f001002.json
│       ...
│       ├── f001010.json
│   └── benthic_tracks
~~~

### Run in Docker

The easiest way to run it is in a docker image. See https://hub.docker.com/u/mbari/deepsea-track for available releases.

To get a list of arguments:

```
docker run mbari/deepsea-track --help
```

## *Arguments* 

                        
  * --video_path absolute path to the video file to process, e.g. /data/video.mp4, /data/video.mov
  * --config_path absolute path to the configuration files: deepsea_class_map.json and deepsea_cfg.json
  * --out_path absolute path to save output artifacts
  * --xml_path (optional) absolute path to directory with voc xml files. If absent, --address and --topic must be set
  * --address (optional) socket address for the detector output, e.g. tcp://127.0.0.1:6432
  * --topic (optional) topic to listen on at address, e.g. VisualEvents. If specified, must also specify the --address 
  * --tracker_width resize width in pixels for running the tracker, defaults to 512
  * --tracker_height resize height in pixels for running the tracker, defaults to 512 
  * --out_width width in pixels to scale the output to, defaults to 1920
  * --out_height height in pixels to scale the output to, defaults to 1080
  * --start_frame_num (optional) starting frame to process, 1-based. e.g. --start_frame=1 is the first frame in the video. Defaults to 1.
  * --stride (optional) amount to stride seeding new detection between frames. A larger stride may process faster. Defaults to 1.
  

## Example

Docker commands:

 * -it = run interactively
 * --rm = remove after execution
 * -v $PWD:/data = mount the current working directory to /data in the container

e.g.
- process video file /data/benthic/video.mp4
- configuration files in /data/benthic
- output results to mapped /data mount in the directory /data/benthic_tracks
- *no start frame specified - start at frame 1*
- *no stride specified - process every frame; only strides detection input*
- *no tracker_width specified - default to 512*
- *no tracker_height specified - default to 512*
- *no out_width specified - default to 1920*
- *no out_height specified - default to 1080*


```
docker run -it --rm -v $PWD:/data mbari/deepsea-track --video_path /data/benthic/video.mp4 --cfg_path /data/benthic/ --xml_path /data/benthic --out_path /data/benthic_tracks/
```

The output will look like:
           
~~~ 
│   └── benthic_tracks
│       ├── f000001.json
│       ├── f000002.json
│       ├── f000003.json
│       ├── f000004.json
│       ...    
│       ├── results.mp4    
~~~

---

## JSON output schema

Data is output per each frame with all events tracked per the following schema:

```json
{
   "$schema": "http://json-schema.org/draft-04/schema#",
   "title": "deepsea-track",
   "description": "a collection of visual events",
   "type": "object",
	
   "properties": {
	
      "uuid": {
         "description": "The unique identifier for this VisualEvent",
         "type": "string"
      },
		
      "bounding_box": {
         "description": "Bounding Box",
         "type": "object",
         "properties": {
            "x": "top left x coordinate",
            "y": "top left y coordinate",
            "width": "The width in pixels from top left",
            "height": "The height in pixels from top left"
        }        
      },
		
      "occlusion": {
         "description": "Occlusion - an approximation based on intersection over union",
         "type": "integer"
      },

      "class_name": {
         "description": "The unique class name ",
         "type": "string"
      },

      "class_index": {
         "description": "The unique class index",
         "type": "integer"
      },

      "confidence": {
         "description": "Confidence score for a given class_name/index between 0-100",
         "type": "integer"
      },

      "surprise": {
         "description": "Surprise factor 0-TBD",
         "type": "integer"
      }, 
   },
	
   "required": ["uuid", "bounding_box", "occlusion",  "class_name", "class_index", "confidence", "surprise"]
}
```

---

### Configuration files

The *deep_class_map.json* should contain the mapping from your IDs to actual class names.
This is human-readable descriptions and colors to associate to each class, e.g.
```json
{
    "items":[
      {
        "id": 1,
        "map": [{"name": "PENIAGONE_VITRAE", "description": "Peniagone vitrea", "color": "89  233 206"}]
      },
      {
        "id": 2,
        "map": [{"name": "PENIAGONE_SP_A", "description": "Peniagone sp. A", "color": "211  199 0"}]
      }
    ]
}
```

The *deepsea_cfg.json* should contain the index for the type of tracker you'd like to use:
These correspond to a few available [OpenCV trackers](https://docs.opencv.org/). 

| id   | Tracker           |  Comments |
|----------|---------------|--------------|
| -1 | No tracker  |  |
| 0 | CSRT Discriminative Correlation Filter with Channel and Spatial Reliability  | Slower, but better performance than KCF |
| 1 | KCF Kernelized Correlation Filter  | Fast, but does not handle scale changes and unpredictable movement well |

e.g. this uses the CSRT tracker:
*deepsea_cfg.json*
```json
{
  "tracker": 0,
  "create_video": true,
  "min_event_frames": 3,
  "display": false,
  "nms_threshold": 0.5,
  "score_threshold": 0.4,
  "display_wait_msecs": 2000,
  "tracker_stride": 2,
  "gamma_enhance": false
}
```

| field   | description           |
|----------|---------------|
| tracker | The tracker id - must be a valid id  |
| tracker_stride | Run the tracker every 2 frames instead of every frame for speed-up |
| gamma_enhance |  Gamma enhance the input video. Brightens dark videos; caution - this add processing time|
| tracker_wait_msecs |  The time in milliseconds to pause between running the tracker on each frame; useful if detector messages are slow over zmq |
| create_video | Set *true* to create a .mp4 video of the output; caution - this adds processing time |
| display | Set *true* to see the frame output displayed while running |
| display_wait_msecs | The time in milliseconds to wait to while displaying the frame output. Useful for debugging. |
| min_event_frames | Only VisualEvents greater than this number of frames are stored; used to remove short invalid tracks. |
| nms_threshold | The minimum score for the non maximum supression algorithm.  If nms_threshold is set too low, e.g. 0.1, it will not detect overlapping objects of same or different classes; set too high e.g. 1, will return multiple boxes for the same object. |
| score_threshold | The minimum score for detections. Set this low to allow more weak detections. |

# Building

An image is available on hub.docker.com at [mbari/deepsea-track](https://hub.docker.com/r/mbari/deepsea-track).
This can be build from source with:

## Docker
```
docker build -t deepsea-track -f Dockerfile.amd .
```

## Mac OSX Native
Note that the CMake files in this project require version 3.15 or higher
```
brew install boost
brew tap nlohmann/json
brew install nlohmann/json
brew install cppzmq
brew install opencv
brew install xerces
```

## Roadmap
-  Add support for inline object detection inference using the [OpenVINO Toolkit](docs.openvinotoolkit.org)
-  Add to docker build 
```
cd ../../thirdparty &&
git clone --recursive https://github.com/Microsoft/onnxruntime && \
    cd ./onnxruntime && build.sh --config RelWithDebInfo --build_shared_lib --parallel && \
    cd ./Linux/RelWithDebInfo && make install lib
```


CHANGELOG.md
# DeepSea Track Changelog

## [1.17.5](http://bitbucket.org/mbari/deepsea-track/compare/v1.17.4...v1.17.5) (2021-12-18)


### Bug Fixes

* correct location for ACK ([114123c](http://bitbucket.org/mbari/deepsea-track/commits/114123cc0e8949fc5b4181c89417495e5dddff12))

## [1.17.4](http://bitbucket.org/mbari/deepsea-track/compare/v1.17.3...v1.17.4) (2021-12-17)


### Performance Improvements

* switch to zmq client/server model, 1-based frame from detector, and release frame if no detections ([ab3fbd6](http://bitbucket.org/mbari/deepsea-track/commits/ab3fbd6ac9e5d3defb54905a096ac29b515ec202))

## [1.17.3](http://bitbucket.org/mbari/deepsea-track/compare/v1.17.2...v1.17.3) (2021-12-17)


### Bug Fixes

* correct zmq frame indexing ([d5a49fc](http://bitbucket.org/mbari/deepsea-track/commits/d5a49fc33549f4d5ef399a988e06f4b9a25a8c79))

## [1.17.2](http://bitbucket.org/mbari/deepsea-track/compare/v1.17.1...v1.17.2) (2021-12-14)


### Bug Fixes

* disconnect subscriber on exit ([eff17f1](http://bitbucket.org/mbari/deepsea-track/commits/eff17f1eb72c3e69bbd7bef1ad3473c746abaa28))

## [1.17.1](http://bitbucket.org/mbari/deepsea-track/compare/v1.17.0...v1.17.1) (2021-12-14)


### Bug Fixes

* correct frame number cast to int and check ([cd64a8d](http://bitbucket.org/mbari/deepsea-track/commits/cd64a8d32078f77dfd7fa19d6405c1adfe49c185))
* remove blocking code ([2bcccd4](http://bitbucket.org/mbari/deepsea-track/commits/2bcccd4ca9ee85ee89ddb80e8a79af0c3f8d8def))


### Performance Improvements

* experimental rework of zmq ([94c4453](http://bitbucket.org/mbari/deepsea-track/commits/94c44533ef66af8a90037218da3ca87bffef5e90))
* removed tracker wait ([1c75d19](http://bitbucket.org/mbari/deepsea-track/commits/1c75d1912b3647003e7d57f0de786ed2209520a1))

# [1.17.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.16.3...v1.17.0) (2021-12-05)


### Features

* initial check-in ([12ec2bb](http://bitbucket.org/mbari/deepsea-track/commits/12ec2bb62430062c81b4f5b8bec1a4142fa78096))


### Performance Improvements

* dynamic template sizing and other minor memory mgmt improvements ([e7ac867](http://bitbucket.org/mbari/deepsea-track/commits/e7ac867a1e2945454cbc5778f719e49b5791414b))
* dynamic template sizing and other minor memory mgmt improvements ([524692f](http://bitbucket.org/mbari/deepsea-track/commits/524692fe9a9bd0eac5e4a363b221417a41f4206f))

## [1.16.3](http://bitbucket.org/mbari/deepsea-track/compare/v1.16.2...v1.16.3) (2021-11-22)


### Performance Improvements

*  added fast tracker wrapper, optional gamma for speed-up, removed unused code, log programs args, removed MEDIANFLOW/TLD/MOSSE trackers. ([63b89e1](http://bitbucket.org/mbari/deepsea-track/commits/63b89e1df15b09a75f36b5919ec9893f25f5da40))

## [1.16.2](http://bitbucket.org/mbari/deepsea-track/compare/v1.16.1...v1.16.2) (2021-11-05)


### Performance Improvements

* skip over loop if no detections, objects and not generating visuals ([617768e](http://bitbucket.org/mbari/deepsea-track/commits/617768e7c6ba425d3df20b8860035f68f2be4463))

## [1.16.1](http://bitbucket.org/mbari/deepsea-track/compare/v1.16.0...v1.16.1) (2021-07-15)


### Performance Improvements

* move listening to outside of loop ([c633c40](http://bitbucket.org/mbari/deepsea-track/commits/c633c400af960df6ac7c878ac2b1db04c4b23849))

# [1.16.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.15.0...v1.16.0) (2021-06-16)


### Features

* tracking improvements ([492e9de](http://bitbucket.org/mbari/deepsea-track/commits/492e9de9b9d4fd50490ef6900f5f2449cda65de4))

# [1.15.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.14.0...v1.15.0) (2021-06-08)


### Bug Fixes

* pass through float instead of TrackerType for score and nms ([adbcbe5](http://bitbucket.org/mbari/deepsea-track/commits/adbcbe59ff2d3ab81ab2062c7ce5302aa974e08a))


### Features

* only short first 7 characters of uuid for brevity and print exact path to cfg ([12ceaf9](http://bitbucket.org/mbari/deepsea-track/commits/12ceaf9beff786917853c1745607dafb0c3f3e29))

# [1.14.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.13.1...v1.14.0) (2021-06-04)


### Features

* added option to create video and prefix with video file stem ([e6fe3ab](http://bitbucket.org/mbari/deepsea-track/commits/e6fe3ab37b1fc8670d993b2c06e6662992b38e67))

## [1.13.1](http://bitbucket.org/mbari/deepsea-track/compare/v1.13.0...v1.13.1) (2021-06-01)


### Bug Fixes

* zmq boxes normalized to 0-1 float ([b713c87](http://bitbucket.org/mbari/deepsea-track/commits/b713c87d7ced29d868291890d2c14560d23a2e3d))

# [1.13.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.12.0...v1.13.0) (2021-06-01)


### Features

* added configuration of nms and score thresholds to tracker ([9286dfc](http://bitbucket.org/mbari/deepsea-track/commits/9286dfc9337e322b66d81b5638086711b4def5d3))

# [1.12.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.11.0...v1.12.0) (2021-06-01)


### Features

* added output scaling and better renaming to distinguish resize ([4145dd0](http://bitbucket.org/mbari/deepsea-track/commits/4145dd03f78a5fb42e200c4059ac228eea3abeb9))

# [1.11.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.10.0...v1.11.0) (2021-05-31)


### Features

* added zmq message tracker delay ([be8a424](http://bitbucket.org/mbari/deepsea-track/commits/be8a4240fa176ea5142d533b4836a1cb3c857e1b))

# [1.10.0](http://bitbucket.org/mbari/deepsea-track/compare/v1.9.3...v1.10.0) (2021-05-28)


### Features

* added tracker delay and delay to zmq initialize wait message block ([89bb3f0](http://bitbucket.org/mbari/deepsea-track/commits/89bb3f07fb7ac84bbfc7688efff9e37744578295))

## [1.9.3](http://bitbucket.org/mbari/deepsea-track/compare/v1.9.2...v1.9.3) (2021-05-26)


### Bug Fixes

* correct warning ([77e1d2e](http://bitbucket.org/mbari/deepsea-track/commits/77e1d2e8cd43c42b2a0f6b2c348242e3c19bc4a6))

## [1.9.2](http://bitbucket.org/mbari/deepsea-track/compare/v1.9.1...v1.9.2) (2021-05-24)


### Bug Fixes


src/arguments.cc
/*
 * Copyright 2020 MBARI
 *
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3.0
 * (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 * http://www.gnu.org/copyleft/lesser.html
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This work would not be possible without the generous support of the
 * David and Lucile Packard Foundation
 */
//!\file Arguments.h  Manages parsing the command line Arguments

#include <iostream>
#include <sys/stat.h>
#include <boost/program_options.hpp>
#include <deepsea/arguments.h>

using namespace std;
namespace po = boost::program_options;

namespace deepsea {

// ######################################################################
    Arguments::Arguments(int argc, char **argv){

        initialized_ = false;
        po::options_description desc("Allowed Arguments");
        desc.add_options()
                ("help", "produce help message")
                ("start_frame_num", po::value<unsigned int>(&start_frame_num_)->default_value(0),
                        "starting frame to process, e.g. --start_frame=0 is the first frame in the video")
                ("tracker_width", po::value<unsigned int>(&tracker_width_)->default_value(512),
                        "resize width in pixels for running the tracker")
                ("tracker_height", po::value<unsigned int>(&tracker_height_)->default_value(512),
                        "resize width in pixels for running the tracker")
                ("out_width", po::value<unsigned int>(&out_width_)->default_value(1920),
                 "width in pixels the JSON output should be scaled to")
                ("out_height", po::value<unsigned int>(&out_height_)->default_value(1080),
                 "width in pixels the JSON output should be scaled to")
                ("stride", po::value<unsigned int>(&stride_)->default_value(1),
                        "stride in detections, 1-based. e.g. --stride=5 will skip ip seeding new events to every 5th detection")
                ("video_path", po::value<string>(&video_path_),
                        "absolute path of the video file to process")
                ("out_path", po::value<string>(&out_path_),
                 "absolute path to save all output to")
                ("cfg_path", po::value<string>(&cfg_path_),
                 "absolute path to the configuration artifacts: deepsea_class_map.json and deepsea_cfg.json files")
                ("xml_path", po::value<string>(&xml_path_),
                 "absolute path to directory with voc xml files. If absent, --address and --topic must be set")
                ("address", po::value<string>(&address_),
                        "socket address for the detector output, e.g. tcp://127.0.0.1:6432")
                ("xml_format", po::value<string>(&xml_format_)->default_value("f%06d.xml"),
                        "format of the xml input files. defaults to f%06d.xml")
                ;

        po::variables_map vm;
        po::store(po::parse_command_line(argc, argv, desc), vm);
        po::notify(vm);
        struct stat buffer;

        // TODO: add a check for a valid starting xml input file here per the xml_format flag

        if (vm.count("help")) {
            cout << desc << "\n";
            return;
        }
        bool failure = this->detection_options(vm);

        if (vm.count("cfg_path")) {
            if (stat (cfg_path_.c_str(), &buffer) != 0) {
                cout << "Directory " << cfg_path_ << " does not exist" << endl;
                failure = true;
            }
            else {
                cout << "Configuration path was set to " << cfg_path_ << endl;
                string filename = cfg_path_ + "deepsea_class_map.json";
                if (stat (filename.c_str(), &buffer) != 0) {
                    cout << filename << " does not exist";
                    failure = true;
                }
                filename = cfg_path_ + "deepsea_cfg.json";
                if (stat (filename.c_str(), &buffer) != 0) {
                    cout << filename << " does not exist";
                    failure = true;
                }
            }
        }
        else {
            cout << "Configuration path must be set with --cfg_path" << endl;
            failure = true;
        }
        if (vm.count("video_path")) {
            string video_path =  this->video_path_;
            if (stat (video_path.c_str(), &buffer) != 0) {
                cout << video_path << " does not exist";
                failure = true;
            }
            else {
                cout << "Video to process was set to " << video_path << endl;
            }
        } else {
            cout << "Must set the video_path with --video_path" << endl;
            failure = true;
        }
        if (vm.count("out_path")) {
            if (stat (out_path_.c_str(), &buffer) != 0) {
                cout << "Directory " << out_path_ << " does not exist" << endl;
                failure = true;
            }
            else {
                cout << "Output path to log files was set to " << out_path_ << endl;
            }
        }
        if (!vm.count("start_frame_num")) {
            start_frame_num_ = 0;
        }
        cout << "Start processing frame " << start_frame_num_ << endl;

        if (vm.count("tracker_width_") && vm.count("tracker_height_")) {
            cout << "Tracking size to " << tracker_width_ << "x" << tracker_height_ << endl;
        }
        if (vm.count("out_width_") && vm.count("out_height_")) {
            cout << "JSON output will be scaled to " << out_width_ << "x" << out_height_ << endl;
        }
        if (vm.count("stride")) {
            cout << "Skipping every " << stride_ << " frame" << endl;
        }
        if (!failure)
            initialized_ = true;
    }

// ######################################################################
    Arguments::~Arguments()
    {
    }

    bool Arguments::detection_options(const boost::program_options::variables_map & vm)
    {
        bool failure = false;
        struct stat buffer;

        if (vm.count("xml_path") && (vm.count("address") || vm.count("topic") ))  {
            throw std::logic_error(std::string("Specify --topic and --address or detection input from --xml_path"));
        }

        if (vm.count("xml_path")) {
            if (stat (xml_path_.c_str(), &buffer) != 0) {
                cout << "Directory " << xml_path_ << " does not exist" << endl;
                failure = true;
            }
            else {
                cout << "Input path with xml detections was set to " << xml_path_ << endl;
            }
        }
        return failure;

    }

}